tools/kagami/src/swarm.rs

```diff
@@ -0,0 +1,1314 @@
+    /// `volumes` field.
+    ///
+    /// If you don't need the defaults, you could set this flag. The `config` directory will be
+    /// created anyway, but you should put the necessary configuration there by yourself.
+    #[arg(long)]
+    no_default_configuration: bool,
+    /// Might be useful for deterministic key generation.
+    ///
+    /// It could be any string. Its UTF-8 bytes will be used as a seed.
+    #[arg(long, short)]
+    seed: Option<String>,
+}
+
+impl Args {
+    pub fn run(self) -> Outcome {
+        let ui = UserInterface::new();
+
+        let prepare_dir_strategy = if self.outdir_force {
+            PrepareDirectoryStrategy::ForceRecreate
+        } else {
+            PrepareDirectoryStrategy::Prompt
+        };
+        let source = ImageSource::from(self.source);
+        let target_dir = TargetDirectory::new(AbsolutePath::absolutize(self.outdir)?);
+
+        if let EarlyEnding::Halt = target_dir
+            .prepare(&prepare_dir_strategy, &ui)
+            .wrap_err("failed to prepare directory")?
+        {
+            return Ok(());
+        }
+
+        let config_dir = AbsolutePath::absolutize(target_dir.path.join(DIR_CONFIG))?;
+
+        let source = source
+            .resolve(&target_dir, &ui)
+            .wrap_err("failed to resolve the source of image")?;
+
+        let ui = if self.no_default_configuration {
+            PrepareConfigurationStrategy::GenerateOnlyDirectory
+        } else {
+            PrepareConfigurationStrategy::GenerateDefault
+        }
+        .run(&config_dir, ui)
+        .wrap_err("failed to prepare configuration")?;
+
+        DockerComposeBuilder {
+            target_dir: target_dir.path.clone(),
+            config_dir,
+            source,
+            peers: self.peers,
+            seed: self.seed.map(String::into_bytes),
+        }
+        .build()
+        .wrap_err("failed to build docker compose")?
+        .write_file(&target_dir.path.join(FILE_COMPOSE))
+        .wrap_err("failed to write compose file")?;
+
+        ui.log_complete(&target_dir.path);
+
+        Ok(())
+    }
+}
+
+#[derive(ClapArgs, Clone, Debug)]
+#[group(required = true, multiple = false)]
+struct ImageSourceArgs {
+    /// Use specified docker image.
+    #[arg(long)]
+    image: Option<String>,
+    /// Use local path location of the Iroha source code to build images from.
+    ///
+    /// If the path is relative, it will be resolved relative to the CWD.
+    #[arg(long, value_name = "PATH")]
+    build: Option<PathBuf>,
+    /// Clone `hyperledger/iroha` repo from the revision Kagami is built itself,
+    /// and use the cloned source code to build images from.
+    #[arg(long)]
+    build_from_github: bool,
+}
+
+/// Parsed version of [`ImageSourceArgs`]
+#[derive(Clone, Debug)]
+enum ImageSource {
+    Image {
+        name: String,
+    },
+    GitHub {
+        revision: String,
+    },
+    /// Raw path passed from user
```

Comment:
I would say it's pretty descriptive without the doc comment, so either adding docs to the remaining variants or ditching all of them sounds as a more consistent choice.

Reply:
Removed